core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -3072,14 +3099,55 @@ private void closeAppendMemoryTruncate(boolean truncate) {
         }
     }
 
-    private void closeWalColumns() {
-        for (int col = 0, n = walMappedColumns.size(); col < n; col++) {
-            MemoryCMOR mappedColumnMem = walMappedColumns.getQuick(col);
-            if (mappedColumnMem != null) {
-                Misc.free(mappedColumnMem);
-                walColumnMemoryPool.push(mappedColumnMem);
+    private void closeWalColumns(boolean isLastSegmentUsage, long walSegmentId) {
+        int key = walFdCache.keyIndex(walSegmentId);
+        if (isLastSegmentUsage || columnCount > configuration.getWalMaxFileDescriptorsCache()) {
+            if (key < 0) {
+                IntList fds = walFdCache.valueAt(key);
+                walFdCache.removeAt(key);
+                walFdCacheSize -= fds.size();
+                fds.clear();
+                walFdCacheListPool.push(fds);
+            }
+
+            for (int col = 0, n = walMappedColumns.size(); col < n; col++) {
+                MemoryCMOR mappedColumnMem = walMappedColumns.getQuick(col);
+                if (mappedColumnMem != null) {
+                    Misc.free(mappedColumnMem);
+                    walColumnMemoryPool.push(mappedColumnMem);
+                }
+            }
+        } else {
+            IntList fds = null;
+            if (key > -1) {
+                // Add FDs to a new FD cache list
+                fds = walFdCacheListPool.pop();
+                walFdCache.putAt(key, walSegmentId, fds);
+            }
+
+            for (int col = 0, n = walMappedColumns.size(); col < n; col++) {
+                MemoryCMOR mappedColumnMem = walMappedColumns.getQuick(col);
+                if (mappedColumnMem != null) {
+                    int fd = mappedColumnMem.detachFdClose();
+                    if (fds != null) {
+                        fds.add(fd);
+                        walFdCacheSize++;
+                    }
+                    walColumnMemoryPool.push(mappedColumnMem);
+                }
             }
         }
+
+        if (walFdCacheSize > configuration.getWalMaxFileDescriptorsCache()) {
+            // Close all cached FDs
+            closeWalFiles();
```

Comment:
Why do we close all files here instead of closing some of them until the size is under `configuration.getWalMaxFileDescriptorsCache()`?

Reply:
It would not be trivial to devise an adequate strategy like LRU. Inadequate strategy would be counterproductive, flushing them is supposed to be a rare occasion.

Reply:
Yes, ideally we need a simple wal/segment id-based LRU cache here or maybe even a FIFO cache if we expect all WAL writers to be relatively equally productive in terms of the number of inserted rows. Looks like this deserves at least a TODO.

Reply:
I changed the configuration to limit the number of segments in this cache, not the number of FDs.

```
# Maximum number of Segments to cache File descriptors for when applying from WAL to table storage.
# WAL segment files are cached to avoid opening and closing them on processing each commit.
# Ideally should be in line with average number of simultaneous connections writing to the tables.
#cairo.wal.max.segment.file.descriptors.cache=30
```

I left a comment that the closing all supposed to be a rare event and not worth optmising for.